--- /root/SP7021/boot/uboot/include/configs/pentagram_common.h	2023-07-12 06:29:09.000000000 +0000
+++ /root/LTPP3_ROOTFS/boot/configs/pentagram_common.h	2023-07-13 05:20:25.000000000 +0000
@@ -117,6 +117,273 @@
 #define RASPBIAN_CMD                    // Enable Raspbian command
 
 
+
+/* Added to handle the situation when the MD-button is pressed */
+/* Note: 'tb_button_state' is determined and set in boot/uboot/board/sunplus/pentagram_board/sp_go.c */
+#define MD_BUTTON_ISP_USB1	0x27
+
+#define MD_BUTTON_VARIABLES_DEFINE \
+"setenv ISP_NULL null \n; " \
+"setenv ISP_IF_SD1 sd_mmc_1 \n; " \
+"setenv ISP_IF_USB0 usb_dev_0 \n; " \
+"setenv ISP_IF_USB1 usb_dev_1 \n; " \
+"setenv SDDEV1_CMD mmc dev 1 \n; " \
+"setenv USBDEV0_CMD usb dev 0 \n; " \
+"setenv USBDEV1_CMD usb dev 1 \n; " \
+"setenv isp_if_test $ISP_NULL \n; " \
+"setenv isp_bootseq1 $ISP_NULL \n; " \
+"setenv isp_bootseq2 $ISP_NULL \n; " \
+"setenv isp_bootseq3 $ISP_NULL \n; " \
+"setenv tb_button_state $ISP_NULL \n; " \
+
+#define MD_BUTTON_VARIABLES_INIT_SHOW \
+"echo ---:tibbo:init: variables \n; " \
+"echo ------: tb_button_state=$tb_button_state \n; " \
+"echo ------: ISP_IF_SD1=$ISP_IF_SD1 \n; " \
+"echo ------: ISP_IF_USB0=$ISP_IF_USB0 \n; " \
+"echo ------: ISP_IF_USB1=$ISP_IF_USB1 \n; " \
+"echo ------: SDDEV1_CMD=$SDDEV1_CMD \n; " \
+"echo ------: USBDEV0_CMD=$USBDEV0_CMD \n; " \
+"echo ------: USBDEV1_CMD=$USBDEV1_CMD \n; " \
+"echo ------: isp_if_test=$isp_if_test \n; " \
+"echo ------: isp_bootseq1=$isp_bootseq1 \n; " \
+"echo ------: isp_bootseq2=$isp_bootseq2 \n; " \
+"echo ------: isp_bootseq3=$isp_bootseq3 \n; " \
+
+#define MD_BUTTON_VARIABLES_ENDRESULT_SHOW \
+"echo ---:tibbo:end-result: variables \n; " \
+"echo ------: tb_button_state=$tb_button_state \n; " \
+"echo ------: ISP_IF_SD1=$ISP_IF_SD1 \n; " \
+"echo ------: ISP_IF_USB0=$ISP_IF_USB0 \n; " \
+"echo ------: ISP_IF_USB1=$ISP_IF_USB1 \n; " \
+"echo ------: SDDEV1_CMD=$SDDEV1_CMD \n; " \
+"echo ------: USBDEV0_CMD=$USBDEV0_CMD \n; " \
+"echo ------: USBDEV1_CMD=$USBDEV1_CMD \n; " \
+"echo ------: isp_if_test=$isp_if_test \n; " \
+"echo ------: isp_bootseq1=$isp_bootseq1 \n; " \
+"echo ------: isp_bootseq2=$isp_bootseq2 \n; " \
+"echo ------: isp_bootseq3=$isp_bootseq3 \n; " \
+
+#define MD_BUTTON_BOOTSEQ_SELECT \
+"echo ---:tibbo:isp-bootseq-start: selection \n; " \
+"echo ------:tibbo:isp-bootseq-note: please ignore warnings (should there be any) \n; " \
+"if test -e mmc 0:9 /.tb_init_bootseq_sdusb0usb1; then \n; " /* START: check if .tb_init_bootseq_sdusb0usb1 is PRESENT in /tb_reserve */ \
+"echo ------:tibbo:bootseq-chosen: SD > USB0 > USB1 \n; " \
+"setenv isp_bootseq1 run md_button_validate_sd_cmd \n; " \
+"setenv isp_bootseq2 run md_button_validate_usb0_cmd \n; " \
+"setenv isp_bootseq3 run md_button_validate_usb1_cmd \n; " \
+"elif test -e mmc 0:9 /.tb_init_bootseq_sdusb1usb0; then \n; " /* ELIF: check if .tb_init_bootseq_sdusb1usb0 is PRESENT in /tb_reserve */ \
+"echo ------:tibbo:bootseq-chosen: SD > USB1 > USB0 \n; " \
+"setenv isp_bootseq1 run md_button_validate_sd_cmd \n; " \
+"setenv isp_bootseq2 run md_button_validate_usb1_cmd \n; " \
+"setenv isp_bootseq3 run md_button_validate_usb0_cmd \n; " \
+"elif test -e mmc 0:9 /.tb_init_bootseq_usb0sdusb1; then \n; " /* ELIF: check if .tb_init_bootseq_usb0sdusb1 is PRESENT in /tb_reserve */ \
+"echo ------:tibbo:bootseq-chosen: USB0 > SD > USB1 \n; " \
+"setenv isp_bootseq1 run md_button_validate_usb0_cmd \n; " \
+"setenv isp_bootseq2 run md_button_validate_sd_cmd \n; " \
+"setenv isp_bootseq3 run md_button_validate_usb1_cmd \n; " \
+"elif test -e mmc 0:9 /.tb_init_bootseq_usb0usb1sd; then \n; " /* ELIF: check if .tb_init_bootseq_usb0usb1sd is PRESENT in /tb_reserve */ \
+"echo ------:tibbo:bootseq-chosen: USB0 > USB1 > SD \n; " \
+"setenv isp_bootseq1 run md_button_validate_usb0_cmd \n; " \
+"setenv isp_bootseq2 run md_button_validate_usb1_cmd \n; " \
+"setenv isp_bootseq3 run md_button_validate_sd_cmd \n; " \
+"elif test -e mmc 0:9 /.tb_init_bootseq_usb1sdusb0; then \n; " /* ELIF: check if .tb_init_bootseq_usb1sdusb0 is PRESENT in /tb_reserve */ \
+"echo ------:tibbo:bootseq-chosen: USB1 > SD > USB0 \n; " \
+"setenv isp_bootseq1 run md_button_validate_usb1_cmd \n; " \
+"setenv isp_bootseq2 run md_button_validate_sd_cmd \n; " \
+"setenv isp_bootseq3 run md_button_validate_usb0_cmd \n; " \
+"elif test -e mmc 0:9 /.tb_init_bootseq_usb1usb0sd; then \n; " /* ELIF: check if .tb_init_bootseq_usb1usb0sd is PRESENT in /tb_reserve */ \
+"echo ------:tibbo:bootseq-chosen: USB1 > USB0 > SD \n; " \
+"setenv isp_bootseq1 run md_button_validate_usb1_cmd \n; " \
+"setenv isp_bootseq2 run md_button_validate_usb0_cmd \n; " \
+"setenv isp_bootseq3 run md_button_validate_sd_cmd \n; " \
+"else \n; " /* ELSE: for all other cases */ \
+"echo ------:tibbo:bootseq-default: SD > USB0 > USB1 \n; " \
+"setenv isp_bootseq1 run md_button_validate_sd_cmd \n; " \
+"setenv isp_bootseq2 run md_button_validate_usb0_cmd \n; " \
+"setenv isp_bootseq3 run md_button_validate_usb1_cmd \n; " \
+"fi \n; " \
+"echo ---:tibbo:isp-bootseq-end: selection \n; " \
+
+#define MD_BUTTON_VALIDATE_SD \
+"echo ---:tibbo:detect: sd-card state... \n; " \
+"$SDDEV1_CMD \n; " \
+"if test $? = 0; then \n; " /* START: check if sd-card is PRESENT */ \
+"if test -e mmc 1:1 /ISPBOOOT.BIN; then \n; " /* START: check if ISPBOOOT.BIN is PRESENT */ \
+"setenv isp_if_test $ISP_IF_SD1 \n; " \
+"else \n; " /* ELSE: check if ISPBOOOT.BIN is PRESENT */ \
+"echo ---:tibbo:sd-card (mmc 1:1): ISPBOOOT.BIN *NOT* found \n; " \
+"fi \n; " /* END: check if ISPBOOOT.BIN is PRESENT */ \
+"fi \n; " /* END: check if sd-card is PRESENT */ \
+
+#define MD_BUTTON_VALIDATE_USB0 \
+"echo ---:tibbo:detect: usb-0 state... \n; " \
+"$USBDEV0_CMD \n; " \
+"if test $? = 0; then \n; " /* START: check if usb-0 is PRESENT */ \
+"if test -e usb 0:1 /ISPBOOOT.BIN; then \n; " /* START: check if ISPBOOOT.BIN is PRESENT */ \
+"setenv isp_if_test $ISP_IF_USB0 \n; " \
+"else \n; " /* ELSE: check if ISPBOOOT.BIN is PRESENT */ \
+"echo ---:tibbo:usb0 (usb 0:1): ISPBOOOT.BIN *NOT* found \n; " \
+"fi \n; " /* END: check if ISPBOOOT.BIN is PRESENT */ \
+"fi \n; " /* END: check if usb-0 is PRESENT */ \
+
+#define MD_BUTTON_VALIDATE_USB1 \
+"echo ---:tibbo:detect: usb-1 state... \n; " \
+"$USBDEV0_CMD \n; " \
+"if test $? = 0; then \n; " /* START: check if usb-0 is PRESENT */ \
+"if test -e usb 1:1 /ISPBOOOT.BIN; then \n; " /* START: check if ISPBOOOT.BIN is PRESENT */ \
+"setenv isp_if_test $ISP_IF_USB1 \n; " \
+"else \n; " /* ELSE: check if ISPBOOOT.BIN is PRESENT */ \
+"echo ---:tibbo:usb1 (usb 1:1): ISPBOOOT.BIN *NOT* found \n; " \
+"fi \n; " /* END: check if ISPBOOOT.BIN is PRESENT */ \
+"fi \n; " /* END: check if usb-0 is PRESENT */ \
+
+#define MD_BUTTON_MEM_WRITE_AND_READ_0X9E809408_0X00000027 \
+"echo ------:tibbo:note: this will run bootcmd 'isp_usb' \n; " \
+"mw.l  0x9e809408  0x00000027 1 \n; " \
+"echo ---:tibbo:get: memory-display of address (0x9e809408)\n; " \
+"md.l 0x9e809408 1 \n; " \
+
+#define MD_BUTTON_MEM_WRITE_AND_READ_0X9E809408_0X00000017 \
+"echo ------:tibbo:note: this will run bootcmd 'isp_usb' \n; " \
+"mw.l  0x9e809408  0x00000017 1 \n; " \
+"echo ---:tibbo:get: memory-display of address (0x9e809408)\n; " \
+"md.l 0x9e809408 1 \n; " \
+
+#define MD_BUTTON_MEM_WRITE_AND_READ_0X9E809408_0X00000007 \
+"echo ---:TIBBO:FORCE-SET: memory-write to address (0x9e809408): 0x00000007 \n; " \
+"echo ------:tibbo:note: this will run bootcmd 'isp_sdcard' \n; " \
+"mw.l  0x9e809408  0x00000007 1 \n; " \
+"echo ---:tibbo:get: memory-display of address (0x9e809408)\n; " \
+"md.l 0x9e809408 1 \n; " \
+
+#define MD_BUTTON_NOTSUPPORTED \
+"echo ---:tibbo:info: md-button feature not enabled... \n; " \
+"echo \n; " \
+
+#define MD_BUTTON_PRINTENV_REMOVE \
+"echo ---:tibbo:printenv: remove 'md_button_sd_or_usb_boot_cmd' \n; " \
+"env delete md_button_sd_or_usb_boot_cmd \n; " \
+"echo ---:tibbo:printenv: remove 'md_button_isp_usb1_cmd' \n; " \
+"env delete md_button_isp_usb1_cmd \n; " \
+"echo ---:tibbo:printenv: remove 'md_button_validate_sd_cmd' \n; " \
+"env delete md_button_validate_sd_cmd \n; " \
+"echo ---:tibbo:printenv: remove 'md_button_validate_usb0_cmd' \n; " \
+"env delete md_button_validate_usb0_cmd \n; " \
+"echo ---:tibbo:printenv: remove 'md_button_validate_usb1_cmd' \n; " \
+"env delete md_button_validate_usb1_cmd \n; " \
+"saveenv \n; " \
+
+#define MD_BUTTON_BOOTCMD_RESET_TO_DEFAULT \
+"echo ---:tibbo:bootcmd: remove all entries \n; " \
+"setenv bootcmd \n; " \
+"echo ---:tibbo:bootcmd: set to default \n; " \
+"setenv bootcmd '" \
+	BOOTCOMMAND_DEFAULT \
+	"'\n; " \
+"saveenv \n; " \
+
+#define BOOTCOMMAND_DEFAULT \
+"echo [scr] bootcmd started; " \
+"md.l ${bootinfo_base} 1; " \
+"if itest.l *${bootinfo_base} == " __stringify(SPI_NOR_BOOT) "; then " \
+	"if itest ${if_zebu} == 1; then " \
+		"if itest ${if_qkboot} == 1; then " \
+			"echo [scr] qk zmem boot; " \
+			"run qk_zmem_boot; " \
+		"else " \
+			"echo [scr] zmem boot; " \
+			"run zmem_boot; " \
+		"fi; " \
+	"else " \
+		"if itest ${if_qkboot} == 1; then " \
+			"echo [scr] qk romter boot; " \
+			"run qk_romter_boot; " \
+		"elif itest.s ${sp_main_storage} == tftp; then " \
+			"echo [scr] tftp_boot; " \
+			"run tftp_boot; " \
+		"else " \
+			"echo [scr] romter boot; " \
+			"run romter_boot; " \
+		"fi; " \
+	"fi; " \
+"elif itest.l *${bootinfo_base} == " __stringify(EMMC_BOOT) "; then " \
+	"if itest ${if_zebu} == 1; then " \
+		"echo [scr] zebu emmc boot; " \
+		"run zebu_emmc_boot; " \
+	"else " \
+		"if itest ${if_qkboot} == 1; then " \
+			"echo [scr] qk emmc boot; " \
+			"run qk_emmc_boot; " \
+		"else " \
+			"echo [scr] emmc boot; " \
+			"run emmc_boot; " \
+		"fi; " \
+	"fi; " \
+"elif itest.l *${bootinfo_base} == " __stringify(SPINAND_BOOT) "; then " \
+	"echo [scr] nand boot; " \
+	"run nand_boot; " \
+"elif itest.l *${bootinfo_base} == " __stringify(USB_ISP) "; then " \
+	"echo [scr] ISP from USB storage; " \
+	"run isp_usb; " \
+"elif itest.l *${bootinfo_base} == " __stringify(SDCARD_ISP) "; then " \
+	"echo [scr] ISP from SD Card; " \
+	"run isp_sdcard; " \
+"else " \
+	"echo Stop; " \
+"fi"
+
+#define MD_BUTTON_RESET \
+"reset \n; " \
+
+#define MD_BUTTON_SD_OR_USB_BOOT \
+MD_BUTTON_VARIABLES_DEFINE \
+"echo \n; " \
+"echo ---:tibbo:detect: md-button state... \n; " \
+"tb_button \n; " \
+"if test $tb_button_state != $ISP_NULL; then \n; " /* START: check if variable is NOT an EMPTY STRING */ \
+	"echo \n; " \
+	"echo **************************************************\n; " \
+	"echo     Boot from sd or usb\n; " \
+	"echo **************************************************\n; " \
+	MD_BUTTON_VARIABLES_INIT_SHOW \
+	"if test $tb_button_state = pressed; then \n; " /* START: check if md-button is PRESSED */ \
+		"echo ---:tibbo:state: md-button is *pressed*... \n; " \
+		MD_BUTTON_BOOTSEQ_SELECT \
+		"echo ---:tibbo:state--------END: md-button is *pressed*... \n; " \
+		"$isp_bootseq1 \n; " /* execute variable as selected in macro 'MD_BUTTON_BOOTSEQ_SELECT' */ \
+		"if test $isp_if_test = $ISP_NULL; then \n; " /* START: for SD: check if isp_if_test is null */ \
+			"$isp_bootseq2 \n; " /* execute variable as selected in macro 'MD_BUTTON_BOOTSEQ_SELECT' */ \
+			"if test $isp_if_test = $ISP_NULL; then \n; " /* START: for USB0: check if isp_if_test is null */ \
+				"$isp_bootseq3 \n; " /* execute variable as selected in macro 'MD_BUTTON_BOOTSEQ_SELECT' */ \
+			"fi \n; " /* END: for USB0: check if isp_if_test is null */ \
+		"fi \n; " /* END: for SD: check if isp_if_test is null */ \
+	"fi \n; " /* END: check if md-button is PRESSED */ \
+	"if test $isp_if_test != $ISP_NULL; then; \n; " /* START: check if isp_if_test is NOT null */ \
+		"if test $isp_if_test = $ISP_IF_USB1; then; \n; " /* START: check if isp_if_test is usb_dev_1 */ \
+			MD_BUTTON_MEM_WRITE_AND_READ_0X9E809408_0X00000027 \
+		"elif test $isp_if_test = $ISP_IF_USB0; then; \n; " /* ELIF: check if isp_if_test is usb_dev_0 */ \
+			MD_BUTTON_MEM_WRITE_AND_READ_0X9E809408_0X00000017 \
+		"else; \n; " /* ELSE: check if isp_if_test is  sd_mmc_1 */ \
+			MD_BUTTON_MEM_WRITE_AND_READ_0X9E809408_0X00000007 \
+		"fi; \n; " /* END: check if isp_if_test is usb_dev_0/mmc_dev_1 */ \
+		MD_BUTTON_VARIABLES_ENDRESULT_SHOW \
+		"else; \n; " /* ELSE: check if isp_if_test is NOT null */ \
+		"echo \n; " \
+		"echo ---:tibbo:state: md-button *not* pressed... \n; " \
+		"echo ---:tibbo:start: boot normally... \n; " \
+	"fi; \n; " /* END: check if isp_if_test is NOT null */ \
+	"echo ************************************************** \n; " \
+	"echo \n; " \
+"else; \n; " /* ELSE: check if variable is NOT an EMPTY STRING */ \
+	MD_BUTTON_NOTSUPPORTED \
+	MD_BUTTON_PRINTENV_REMOVE \
+	MD_BUTTON_BOOTCMD_RESET_TO_DEFAULT \
+	MD_BUTTON_RESET \
+"fi \n; " /* END: check if variable is NOT an EMPTY STRING */ \
+/* Added to handle the situation when the MD-button is pressed */
+
+
 /*
  * In the beginning, bootcmd will check bootmode in SRAM and the flag
  * if_zebu to choose different boot flow :
@@ -154,6 +421,7 @@
  */
 #define CONFIG_BOOTCOMMAND \
 "echo [scr] bootcmd started; " \
+"run md_button_sd_or_usb_boot_cmd; " /* Added to handle the situation when the MD-button is pressed */ \
 "md.l ${bootinfo_base} 1; " \
 "if itest.l *${bootinfo_base} == " __stringify(SPI_NOR_BOOT) "; then " \
 	"if itest ${if_zebu} == 1; then " \
@@ -195,6 +463,9 @@
 "elif itest.l *${bootinfo_base} == " __stringify(USB_ISP) "; then " \
 	"echo [scr] ISP from USB storage; " \
 	"run isp_usb; " \
+"elif itest.l *${bootinfo_base} == " __stringify(MD_BUTTON_ISP_USB1) "; then " \
+	"echo [scr] ISP from USB storage; " \
+	"run md_button_isp_usb1_cmd; " \
 "elif itest.l *${bootinfo_base} == " __stringify(SDCARD_ISP) "; then " \
 	"echo [scr] ISP from SD Card; " \
 	"run isp_sdcard; " \
@@ -364,7 +635,7 @@
 	"cp.l ${addr_src_kernel} ${addr_dst_kernel} ${sz_kernel}; " \
 	dbg_scr("echo sp_go ${addr_dst_kernel} ${fdtcontroladdr}; ") \
 	"sp_go ${addr_dst_kernel} ${fdtcontroladdr}\0" \
-"emmc_boot=sp_wdt_set;" \
+"emmc_boot= sp_wdt_set;" \
 	DTS_LOAD_EMMC \
 	"mmc read ${addr_tmp_header} ${addr_src_kernel} 0x1; " \
 	"setenv tmpval 0; setexpr tmpaddr ${addr_tmp_header} + 0x0c; run be2le; " \
@@ -464,8 +735,26 @@
 	"dhcp $isp_ram_addr $serverip:TFTP0000.BIN; " \
 	"setenv isp_main_storage ${sp_main_storage} && printenv isp_main_storage; " \
 	"setexpr script_addr $isp_ram_addr + 0x00 && setenv script_addr 0x${script_addr} && source $script_addr; " \
+	"\0" \
+"md_button_sd_or_usb_boot_cmd=;" /* Added to handle the situation when the MD-button is pressed */ \
+	MD_BUTTON_SD_OR_USB_BOOT \
+	"\0" \
+"md_button_isp_usb1_cmd=setenv isp_if usb && setenv isp_dev 1; " /* Added to handle the situation when the MD-button is pressed */ \
+	"$isp_if start; " \
+	"run isp_common; " \
+	"\0" \
+"md_button_validate_sd_cmd=;" /* Added to handle the situation when the MD-button is pressed */ \
+	MD_BUTTON_VALIDATE_SD \
+	"\0" \
+"md_button_validate_usb0_cmd=;" /* Added to handle the situation when the MD-button is pressed */ \
+	MD_BUTTON_VALIDATE_USB0 \
+	"\0" \
+"md_button_validate_usb1_cmd=;" /* Added to handle the situation when the MD-button is pressed */ \
+	MD_BUTTON_VALIDATE_USB1 \
 	"\0"
 
+
+
 /* MMC related configs */
 #define CONFIG_SUPPORT_EMMC_BOOT
 /* #define CONFIG_MMC_TRACE */
