--- /root/SP7021/boot/uboot/include/configs/pentagram_common.h	2023-04-13 16:17:25.000000000 +0000
+++ /root/LTPP3_ROOTFS/boot/configs/pentagram_common.h	2023-05-22 10:23:24.000000000 +0000
@@ -117,6 +117,85 @@
 #define RASPBIAN_CMD                    // Enable Raspbian command
 
 
+
+/* Added to handle the situation when the MD-button is pressed */
+/* Note: 'tb_button_state' is determined and set in boot/uboot/board/sunplus/pentagram_board/sp_go.c */
+#define MD_BUTTON_DETECT__SDCARD_OR_USB_BOOT \
+"echo \n; " \
+"echo **************************************************\n; " \
+"echo     BOOT FROM SD OR USB (IF PRESENT)\n; " \
+"echo **************************************************\n; " \
+"echo ---:TIBBO:INITIALIZE: VARIABLES \n; " \
+"setenv tb_button_state released \n; " \
+"setenv ISP_IF_NULL null \n; " \
+"setenv ISP_IF_SD1 sd_mmc_1 \n; " \
+"setenv ISP_IF_USB0 usb_dev_0 \n; " \
+"setenv SDDEV1_CMD mmc dev 1 \n; " \
+"setenv USBDEV0_CMD usb dev 0 \n; " \
+"setenv isp_if_test $ISP_IF_NULL \n; " \
+"echo ------:TIBBO:INIT: tb_button_state=$tb_button_state \n; " \
+"echo ------:TIBBO:INIT: ISP_IF_NULL=$ISP_IF_NULL \n; " \
+"echo ------:TIBBO:INIT: ISP_IF_SD1=$ISP_IF_SD1 \n; " \
+"echo ------:TIBBO:INIT: ISP_IF_USB0=$ISP_IF_USB0 \n; " \
+"echo ------:TIBBO:INIT: SDDEV1_CMD=$SDDEV1_CMD \n; " \
+"echo ------:TIBBO:INIT: USBDEV0_CMD=$USBDEV0_CMD \n; " \
+"echo ------:TIBBO:INIT: isp_if_test=$isp_if_test \n; " \
+"echo \n; " \
+"echo ---:TIBBO:DETECT: IS MD-BUTTON PRESSED...? \n; " \
+"tb_button \n; " \
+"if test -n $tb_button_state; then \n; " /* START: check if variable is NOT an EMPTY STRING */ \
+"echo \n; " \
+"echo ---:TIBBO:RESULT: 'tb_button_state'=$tb_button_state \n; " \
+"if test $tb_button_state = pressed; then \n; " /* START: check if md-button is PRESSED */ \
+"echo \n; " \
+"echo ---:TIBBO:RESULT: MD-BUTTON IS *PRESSED*... \n; " \
+"echo ---:TIBBO:DETECT: IS SD-CARD PRESENT...? \n; " \
+"$SDDEV1_CMD \n; " \
+"if test $? = 0; then \n; " /* START: check if sd-card is PRESENT */ \
+"setenv isp_if_test $ISP_IF_SD1 \n; " \
+"fi \n; " /* END: check if sd-card is PRESENT */ \
+"if test $isp_if_test = $ISP_IF_NULL; then \n; " /* START: check if isp_if_test is null */ \
+"echo \n; " \
+"echo ---:TIBBO:DETECT: IS USB-0 PRESENT...? \n; " \
+"echo ------:TIBBO:NOTE: IF USB-0 AND/OR USB-1 ARE INSERTED, THEN... \n; " \
+"echo ------:TIBBO:NOTE: ...ONLY USB-0 IS CHECKED FOR ITS PRESENCE!\n; " \
+"$USBDEV0_CMD \n; " \
+"if test $? = 0; then \n; " /* START: check if usb-0 is PRESENT */ \
+"setenv isp_if_test $ISP_IF_USB0 \n; " \
+"fi \n; " /* END: check if usb-0 is PRESENT */ \
+"fi \n; " /* END: check if isp_if_test is null */ \
+"fi \n; " /* END: check if md-button is PRESSED */ \
+"else; \n; " /* ELSE: check if variable is NOT an EMPTY STRING */ \
+"echo ---:TIBBO:RESULT: 'tb_button_state' is an *EMPTY STRING* \n; " \
+"fi \n; " /* END: check if variable is NOT an EMPTY STRING */ \
+"if test $isp_if_test != $ISP_IF_NULL; then; \n; " /* START: check if isp_if_test is NOT null */ \
+"if test $isp_if_test = $ISP_IF_USB0; then; \n; " /* START: check if isp_if_test is usb_dev_0 */ \
+"echo \n; " \
+"echo ---:TIBBO:RUN: ISPBOOOT.BIN FROM USB-DEV-0 \n; " \
+"echo ************************************************** \n; " \
+"echo \n; " \
+"run isp_usb; \n; " \
+"else; \n; " /* ELSE: check if isp_if_test is mmc_dev_1 */ \
+"echo \n; " \
+"echo ---:TIBBO:RUN: ISPBOOOT.BIN FROM MMC-DEV-1 \n; " \
+"echo ************************************************** \n; " \
+"echo \n; " \
+"run isp_sdcard; \n; " \
+"fi; \n; " /* END: check if isp_if_test is usb_dev_0 */ \
+"echo \n; " \
+"echo ---:TIBBO:FORCE-SET: memory-write to address (0x9e809408): 0x00000007 \n; " \
+"echo ------:TIBBO:NOTE: this will forcely connect both jumpers (CN10 and CN11) \n; " /* This is necessary to RE-INITIALIZE the U-BOOT partitions */ \
+"mw.l  0x9e809408  0x00000007 1 \n; " \
+"else; \n; " \
+"echo \n; " \
+"echo ---:TIBBO:RESULT: MD-BUTTON IS *NOT PRESSED*... \n; " \
+"echo ---:TIBBO:START: *DEFAULT* BOOTCMD... \n; " \
+"echo ************************************************** \n; " \
+"echo \n; " \
+"fi; \n; " /* END: check if isp_if_test is NOT null */ \
+/* Added to handle the situation when the MD-button is pressed */
+
+
 /*
  * In the beginning, bootcmd will check bootmode in SRAM and the flag
  * if_zebu to choose different boot flow :
@@ -154,6 +233,7 @@
  */
 #define CONFIG_BOOTCOMMAND \
 "echo [scr] bootcmd started; " \
+"run md_button_detect__sdcard_or_usb_boot; " /* Added to handle the situation when the MD-button is pressed */ \
 "md.l ${bootinfo_base} 1; " \
 "if itest.l *${bootinfo_base} == " __stringify(SPI_NOR_BOOT) "; then " \
 	"if itest ${if_zebu} == 1; then " \
@@ -364,7 +444,7 @@
 	"cp.l ${addr_src_kernel} ${addr_dst_kernel} ${sz_kernel}; " \
 	dbg_scr("echo sp_go ${addr_dst_kernel} ${fdtcontroladdr}; ") \
 	"sp_go ${addr_dst_kernel} ${fdtcontroladdr}\0" \
-"emmc_boot=sp_wdt_set;" \
+"emmc_boot= sp_wdt_set;" \
 	DTS_LOAD_EMMC \
 	"mmc read ${addr_tmp_header} ${addr_src_kernel} 0x1; " \
 	"setenv tmpval 0; setexpr tmpaddr ${addr_tmp_header} + 0x0c; run be2le; " \
@@ -464,8 +544,13 @@
 	"dhcp $isp_ram_addr $serverip:TFTP0000.BIN; " \
 	"setenv isp_main_storage ${sp_main_storage} && printenv isp_main_storage; " \
 	"setexpr script_addr $isp_ram_addr + 0x00 && setenv script_addr 0x${script_addr} && source $script_addr; " \
+	"\0" \
+"md_button_detect__sdcard_or_usb_boot=;" /* Added to handle the situation when the MD-button is pressed */ \
+	MD_BUTTON_DETECT__SDCARD_OR_USB_BOOT \
 	"\0"
 
+
+
 /* MMC related configs */
 #define CONFIG_SUPPORT_EMMC_BOOT
 /* #define CONFIG_MMC_TRACE */
