--- /root/SP7021/linux/kernel/drivers/irqchip/irq-sp7021-intc.c	2023-02-17 08:04:10.000000000 +0000
+++ /root/LTPP3_ROOTFS/kernel/drivers/irqchip/irq-sp7021-intc.c	2023-02-17 07:49:25.000000000 +0000
@@ -98,14 +98,17 @@
 #ifdef WORKAROUND_FOR_EDGE_TRIGGER_BUG
 	if (edge_trigger[data->hwirq] & IRQ_TYPE_EDGE_BOTH) {
 		u32 trig_lvl = readl_relaxed(&sp_intc.g0->intr_polarity[idx]);
-// FIX ?
-		if (edge_trigger[data->hwirq] == IRQ_TYPE_EDGE_RISING)
+// FIX FOR THE INTERRUPT ISSUE
+		if (edge_trigger[data->hwirq] == IRQ_TYPE_EDGE_RISING) {
 			trig_lvl |= mask;
-		else
+			edge_trigger[data->hwirq] = IRQ_TYPE_EDGE_FALLING;
+		}
+		else {
 			trig_lvl &= ~mask;
-
+			edge_trigger[data->hwirq] = IRQ_TYPE_EDGE_RISING;
+		}
 		writel_relaxed(trig_lvl, &sp_intc.g0->intr_polarity[idx]);
-		edge_trigger[data->hwirq] |= SP_IRQ_TYPE_EDGE_ACTIVE;
+		
 	}
 #endif
 	writel_relaxed(mask, &sp_intc.g1->intr_clr[idx]);
